### ISSUE TITLE
Incorrect behavior of RemoveAlwaysTrueIfConditionRector

### ISSUE THREAD

**Issue by Bl00D4NGEL** [rectorphp/rector#7850]
# Bug Report

| Subject        | Details              |
| :------------- | :--------------------|
| Rector version | last dev-main      |
| Installed as   | composer dependency  |

## Minimal PHP Code Causing Issue

See https://getrector.com/demo/dad09538-58a1-4b17-9364-0906d49d03fe

```php
<?php

final class Foo {}

final class Bar {}

function testIt((Foo&Bar)|null $fooBar = null) {
	if ($fooBar === null) {
        $fooBar = new Foo();
    }
    
    return $fooBar;
}
```

### Responsible rules

* `RemoveAlwaysTrueIfConditionRector`

## Expected Behavior

No changes

**Comment by staabm** [comment]
Since Foo and Bar are final, I guess the intersection of both will always be `null`. 

I think you example - as is - is simplified too much

https://phpstan.org/r/fbc788a3-c417-4d39-a86d-62707d4c8387

**Comment by samsonasik** [comment]
@staabm so it is not a bug? `$fooBar` always got `NullType`

**Comment by staabm** [comment]
I think so. But maybe the reporter has a better example then what was posted

**Comment by Bl00D4NGEL** [comment]
First of all: Thank you for pointing that out @staabm :)

TL;DR: My bad here, I overlooked some stuff

I looked into this some more and looked at our code again to see where I found this weird behaviour. The "impossible" state that I created as an example in the first place was because of a previous rector run that converted the following block:
```php
/**
 @param Foo&MockObject $foo
*/
public function testIt(?Foo $foo) {}
```

into 
```php
/**
 * @param Foo&MockObject|null $foo
 */
public function testIt((MockObject&Foo)|null $foo) {}
```

Which is a special case here as this was defined in a PHPUnit TestCase where the MockObject can be used as an intersection type with other classes. This still wouldn't allow final classes to be mocked though but for that we use the [BypassFinals](https://github.com/dg/bypass-finals) library which does enable us to mock final classes, rector doesn't pick up on that which is absolutely fair imo.

Therefore I think I just didn't completely analyse the "problem", rector is working just fine here :)
